lightning/src/blinded_path/payment.rs

```diff
@@ -97,6 +100,28 @@ pub struct PaymentConstraints {
 	pub htlc_minimum_msat: u64,
 }
 
+/// The context of an inbound payment, which is included in a [`BlindedPath`] via [`ReceiveTlvs`]
+/// and surfaced in [`Event::PaymentClaimable`].
+///
+/// [`BlindedPath`]: crate::blinded_path::BlindedPath
+/// [`Event::PaymentClaimable`]: crate::events::Event::PaymentClaimable
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum PaymentContext {
+	/// The payment was made for an invoice requested from a BOLT 12 [`Offer`].
+	///
+	/// [`Offer`]: crate::offers::offer::Offer
+	Bolt12Offer {
+		/// The identifier of the [`Offer`].
+		///
+		/// [`Offer`]: crate::offers::offer::Offer
+		offer_id: OfferId,
+	},
+	/// The payment was made for invoice sent for a BOLT 12 [`Refund`].
+	///
+	/// [`Refund`]: crate::offers::refund::Refund
+	Bolt12Refund {},
```

Comment:
Where does the offer id come from? Also is there no id for a refund?

Our main concern for Mutiny is we can't label the receives from bolt 12 if we can't tie it to our original bolt12

Reply:
At the moment, the `OfferId` is a nonce taken from the offer metadata. You can use `OfferBuilder::offer_id` to retrieve it. Maybe we could have a `VerifiedOffer` so you don't need to grab it from the builder before creating the offer.

Note that `OfferId` is the id for inbound payments. We're going to provide most if not all of the data from the `InvoiceRequest` in the `PaymentContext::Bolt12Offer` variant, too. And you'll have the `PaymentHash` from the `Bolt12Invoice` directly in `Event::PaymentClaimable`. So you should be able to use that to correlated all inbound payments to an offer.

For outbound payments, we already use the `PaymentId` passed to `pay_for_offer` which is given back in `Event::PaymentSent`.

For a `Refund` creator (i.e., an outbound payment), the id is the `PaymentId` passed to `create_refund_builder`, which is encrypted in the `payer_metadata` and also given back in `Event::PaymentSent`. For the inbound payment, the id is the `PaymentHash` in the `Bolt12Invoice` sent to the `Refund` creator. That invoice is now returned by `request_refund_payment` in this PR.

Reply:
> I'm still not entirely convinced just using `payment_hash` will be an adequate identifier for inbound payments going forward, given that we're about to potentially have (more) spurious payment events after #2957. It would be really nice if we could also expose `PaymentId` for inbound payments that could function as a "real" idempotency token with documented guarantees for the user, even if we'd derive it from the `PaymentHash` internally. This would make the API more uniform and probably less confusing for the user.

Do you have a concrete proposal in mind?
 
> Also, with BOLT12 multiple invoices may be generated for each offer and each of these could potentially get paid multiple times. Even though the latter is a bug on the payer's end it would be nice if we could expose a way to discern duplicate payments from spurious/duplicate events.

Just to be clear, don't you already need to handle duplicate payments for the same invoice in BOLT 11? And, at least according to our docs, you shouldn't claim duplicate payments: https://docs.rs/lightning/latest/lightning/events/enum.Event.html#note-1